api/helpers/lock-game.js

```diff
@@ -0,0 +1,63 @@
+const { randomUUID } = require('crypto'); // Added in: node v14.17.0
+const dayjs = require('dayjs');
+
+async function sleep(durationInMillis) {
+  return new Promise((resolve) => {
+    setTimeout(resolve, durationInMillis);
+  });
+}
+
+module.exports = {
+  friendlyName: 'Lock a game',
+
+  description:
+    'Lock a game to reserve it for updates until the game is unlocked or the lock times out. Will error if we cannot obtain a lock before hitting the max wait time or the max number of attempts',
+
+  inputs: {
+    gameId: {
+      type: 'number',
+      descritiption: 'The database id of game to lock',
+      example: 44,
+      required: true,
+    },
+  },
+
+  fn: async ({ gameId }, exits) => {
+    const LOCK_RETRY_TIME_MS = 250;
+    const LOCK_MAX_WAIT_TIME_MS = 5000;
+    const MAX_ATTEMPTS = 20;
+    const uuId = randomUUID();
+    const startTime = dayjs();
+    const timeToGiveUp = startTime.add(LOCK_MAX_WAIT_TIME_MS, 'millisecond').valueOf();
+    let now = startTime.valueOf();
+    let numAttempts = 0;
+    while (numAttempts < MAX_ATTEMPTS && now < timeToGiveUp) {
+      try {
+        numAttempts++;
+        now = dayjs().valueOf();
+        const lockIsStaleTimeout = dayjs().subtract(30, 'second').valueOf();
+
+        // Lock & re-fetch game if unlocked or lock is expired
+        const updatedGame = Game.updateOne({
+          id: gameId,
+          or: [
+            {lock: null},
+            {lockedAt: {'<=': lockIsStaleTimeout}}
+          ],
+        }).set({lock: uuId, lockedAt: now});
+
+        // If we successfully wrote our uuid, resolve
+        const newLock = updatedGame?.lock;
```

Comment:
It's notable that sails' `.updateOne()` method automatically queries for the updated record and returns undefined if it's unavailable (as opposed to erroring). That's why we need to optionally chain `updatedGame?.lock`. I considered putting a comment there to explain that behavior, but it seemed like it would clutter the flow of comments about the logic of the helper itself which is why I omitted it. I'm hoping that the update call followed by optionally chaining is relatively clear by itself.